### PR TITLE
K-OPLS implementation

### PR DESCRIPTION
**Pull request by JeffersonH44**

### CODE REVIEW

File: src/__tests__/kopls/kopls.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        orthogonalComponents: 10,
+        predictiveComponents: 1,
+        kernel: kernel
+    });
+
+    cls.train(Xtrain, Ytrain);
+
+    test('K-OPLS test with main features', () => {
+        expect(() => cls.getOrthogonalScoreVectors()).toThrowError(Error);
+        expect(() => cls.getPredictiveScoreMatrix()).toThrowError(Error);
```
Comment (targos) [inline src/__tests__/kopls/kopls.js]: can you assert the error message instead of the constructor?

File: src/__tests__/kopls/kopls.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        var output = cls.predict(Xtest);
+
+        var testTp = cls.getPredictiveScoreMatrix();
+        for (var i = 0; i < testTp.length; ++i) {
+            for (var j = 0; j < testTp[i].length; ++j) {
+                expect(testTp[i][j][0]).toBeCloseTo(Tp[i][j], 2);
```
Comment (targos) [inline src/__tests__/kopls/kopls.js]: @maasencioh made a plugin to simplify this (something like `deepCloseTo()`)

File: package.json [hunk 3]
```diff
@@ -32,13 +37,19 @@
...
     "benchmark": "^2.1.0",
-    "mocha": "^3.0.2",
-    "mocha-better-spec-reporter": "^3.0.2",
+    "eslint": "^3.1.0",
+    "eslint-config-cheminfo": "^1.0.0",
+    "eslint-plugin-no-only-tests": "^1.1.0",
+    "jest": "^20.0.4",
+    "npm-run-all": "^4.0.2",
+    "rollup": "^0.43.0",
     "should": "^11.1.0"
```
Comment (targos) [inline package.json]: do you still need should?

File: README.md [hunk 4]
```diff
@@ -1,25 +1,25 @@
...
   [![npm download][download-image]][download-url]
 
-PLS regression algorithm based on the Yi Cao Matlab implementation:
+PLS regression algorithm based on the Yi Cao implementation:
 
-[Partial Least-Squares and Discriminant Analysis](http://www.mathworks.com/matlabcentral/fileexchange/18760-partial-least-squares-and-discriminant-analysis)
+[PLS Matlab code](http://www.mathworks.com/matlabcentral/fileexchange/18760-partial-least-squares-and-discriminant-analysis)
 
-## installation
+K-OPLS regression algorithm based on [this](http://onlinelibrary.wiley.com/doi/10.1002/cem.1071/abstract) paper, Matlab implementation here:
```
Comment (maasencioh) [inline README.md]: I would put it as `[this paper](...), and the [Matlab implementation](...)`

File: README.md [hunk 5]
```diff
@@ -1,25 +1,25 @@
...
-### new PLS(X, Y)
+`$ npm install ml-pls`
 
-### pls.train(options)
+## Usage
 
-__Example__
+### [PLS](./src/pls.js)
 
 ```js
```
Comment (maasencioh) [inline README.md]: maybe add how do you import it

File: src/kopls.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+            var p3 = p2.mmul(this.to[i]).mmul(toiPrime);
+
+            KteTr[i + 1][i + 1] = p1.sub(p2).add(p3);
+        }
+
+        Tp[i] = KteTr[i][0].mmul(this.Up).mmul(this.SpPow);
+        var prediction = Tp[i].mmul(this.Bt[i]).mmul(this.Cp.transpose());
+
+        this.predScoreMat = Tp;
+        this.predYOrthVectors = to;
```
Comment (targos) [inline src/kopls.js]: I don't really like this approach. Making a prediction shouldn't change the state of the predictor

File: package.json (hunk 3 above)
Comment (maasencioh) [inline package.json]: do you still need this one?

File: src/__tests__/kopls/kopls.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    var cls = new KOPLS({
+        orthogonalComponents: 10,
+        predictiveComponents: 1,
+        kernel: kernel
+    });
+
+    cls.train(Xtrain, Ytrain);
+
+    test('K-OPLS test with main features', () => {
+        expect(() => cls.getOrthogonalScoreVectors()).toThrowError(Error);
```
Comment (maasencioh) [inline src/__tests__/kopls/kopls.js]: is better to put the error message

File: src/kopls.js [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+        this.Cp = Cp;
+        this.SpPow = SpPow;
+        this.Up = Up;
+        this.Tp = Tp;
+        this.co = co;
+        this.so = so;
+        this.to = to;
+        this.toNorm = toNorm;
+        this.Bt = Bt;
+        this.K = KX;
```
Comment (maasencioh) [inline src/kopls.js]: Is very difficult to follow the code if the variable names are like co, so, Tp, etc. Could you please change them to be a little bit more readable?

File: src/kopls.js [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    }
+
+    /**
+     * Get the predictive score matrix for all generations according to the number of orthogonal vectors.
+     * (this can be obtained only after a prediction)
+     * @return {Matrix}
+     */
+    getPredictiveScoreMatrix() {
+        if (!this.predScoreMat) {
+            throw new Error('you should run a prediction first!');
```
Comment (maasencioh) [inline src/kopls.js]: Don't yell at the user!!!!

File: src/opls.js [hunk 10]
```diff
@@ -1,6 +1,5 @@
-'use strict';
 
-var Matrix = require('ml-matrix').Matrix;
+/*var Matrix = require('ml-matrix').Matrix;
```
Comment (maasencioh) [inline src/opls.js]: if the code is not usable is better to move it to a different branch and deleted from here

File: src/__tests__/kopls/kopls.js (hunk 2 above)
Comment (maasencioh) [reply src/__tests__/kopls/kopls.js]: https://github.com/maasencioh/jest-matcher-deep-close-to

File: src/kopls.js (hunk 6 above)
Comment (JeffersonH44) [reply src/kopls.js]: so, should I return those values inside of the return statement?

**Comment by JeffersonH44** [comment]
it's ok @maasencioh @targos?

**Pull request merged**